packages/flutter/lib/src/painting/borders.dart

```diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
   Path getInnerPath(Rect rect, { TextDirection? textDirection });
 
+  /// Paint a canvas with appropriate shape.
+  ///
+  /// This works as an optimization for [getOuterPath] + [Canvas.drawPath],
+  /// which can be slow. Drawing can be faster by providing a [paintInterior]
+  /// with a specific draw method (instead of a generic [Canvas.drawPath]).
+  /// Examples of specific methods are [Canvas.drawRect] and [Canvas.drawCircle].
+  ///
+  /// A standard paintBox implementation is `canvas.draw` followed by
+  /// what is inside [getOuterPath].
+  ///
+  /// As an example, see how [Path.addRect] became [Canvas.drawRect]:
+  ///
+  /// ```dart
+  /// @override
+  /// Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
+  ///  return Path()
+  ///    ..addRect(rect);
+  /// }
+  ///
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) {
+  ///  canvas.drawRect(rect, paint);
+  /// }
+  /// ```
+  ///
+  /// When a shape can only be drawn using path, [preferPaintInterior] is recommended
+  /// to be false, so [ShapeBorder] caches the path.
+  ///
+  /// This function provides a default empty implementation as it is only executed
+  /// when [preferPaintInterior] is true.
+  void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) { }
+
+  /// Decide if [ShapeDecoration] should paint via [paintInterior].
+  ///
+  /// A [ShapeBorder] can paint using [Canvas.drawRect], [Canvas.drawCircle],
+  /// [Canvas.drawPath], etc..
+  /// [Canvas.drawPath] is signifcantly slower than the other draw methods,
+  /// so [ShapeBorder] caches it. In order to cache, it needs to know
+  /// if [paintInterior] calls [Canvas.drawPath] (the slower method) or
+  /// the faster methods.
+  ///
+  /// Caching a circle will lead to reduced performance,
+  /// so [preferPaintInterior] allows the choice.
+  /// When true, ShapeBorder will reduce caching and call [paintInterior].
+  /// When false, ShapeBorder will cache the path first.
+  ///
+  /// GOOD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawRect(rect, paint);
+  /// }
+  ///
+  /// @override
+  /// bool get preferPaintInterior => true;
+  /// ```
+  ///
+  /// GOOD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawPath(_getPath(rect), paint);
+  /// }
+  /// @override
+  /// bool get preferPaintInterior => false;
+  /// ```
+  ///
+  /// BAD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawPath(_getPath(rect), paint);
+  /// }
+  /// @override
+  /// bool get preferPaintInterior => true;
+  /// ```
+  bool get preferPaintInterior => false;
```

Comment:
Tried to rewrite this to be more succint:

```suggestion
  /// Reports whether [paintInterior] is implemented.
  ///
  /// Classes such as [ShapeDecoration] prefer to use [paintInterior] if this
  /// getter returns true. This is intended to enable faster painting; instead of
  /// computing a shape using [getOuterPath] and then drawing it using
  /// [Canvas.drawPath], the path can be drawn directly to the [Canvas] using
  /// dedicated methods such as [Canvas.drawRect] or [Canvas.drawCircle].
  ///
  /// By default, this getter returns false.
  ///
  /// Subclasses that implement [paintInterior] should override this to return true.
  /// Subclasses should only override [paintInterior] if doing so enables faster
  /// rendering than is possible with [Canvas.drawPath] (so, in particular, 
  /// subclasses should not call [Canvas.drawPath] in [paintInterior]).
  ///
  /// See also:
  ///
  ///  * [paintInterior], whose API documentation has an example implementation.
  bool get preferPaintInterior => false;
```